src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/QualityScoreDistributionSpark.java

```diff
@@ -0,0 +1,181 @@
+import org.broadinstitute.hellbender.utils.io.Resource;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+
+import java.io.*;
+import java.util.List;
+
+/**
+ * Charts quality score distribution within a BAM file.
+ *
+ * This is the spark version of the tool.
+ */
+@CommandLineProgramProperties(
+        summary = "Program to chart " +
+                "quality score distributions in a SAM or BAM file.",
+        oneLineSummary = "Charts quality score distributions for a SAM or BAM file",
+        programGroup = SparkProgramGroup.class
+)
+public final class QualityScoreDistributionSpark extends GATKSparkTool {
+    private static final long serialVersionUID = 1l;
+
+    @Argument(doc = "uri for the output file: a local file path",
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            optional = true)
+    public String out;
+
+    @Argument(shortName="CHART", doc="A file (with .pdf extension) to write the chart to.", optional=true)
+    public File CHART_OUTPUT;
+
+    @Argument(doc="If set to true calculate mean quality over aligned reads only.")
+    public boolean ALIGNED_READS_ONLY = false;
+
+    @Argument(shortName="PF", doc="If set to true calculate mean quality over PF reads only.")
+    public boolean PF_READS_ONLY = false;
+
+    @Argument(doc="If set to true, include quality for no-call bases in the distribution.")
+    public boolean INCLUDE_NO_CALLS = false;
+
+    @VisibleForTesting
+    static final class Counts implements Serializable{
+        private static final long serialVersionUID = 1L;
+        private final long[] qCounts;
+        private final long[] oqCounts;
+        private final boolean includeNoCalls;
+
+        Counts(final boolean includeNoCalls){
+           qCounts = new long[128];
+           oqCounts = new long[128];
+           this.includeNoCalls = includeNoCalls;
+        }
+
+        /**
+         * Adds a read to this count object by increasing counts of the base qualities.
+         */
+        Counts addRead(final GATKRead read) {
+            final byte[] bases = read.getBases();
+            final byte[] quals = read.getBaseQualities();
+            final byte[] oq    = ReadUtils.getOriginalBaseQualities(read);
+
+            final int length = quals.length;
+
+            for (int i=0; i<length; ++i) {
+                if (includeNoCalls || !SequenceUtil.isNoCall(bases[i])) {
+                    qCounts[quals[i]]++;
+                    if (oq != null) {
+                        oqCounts[oq[i]]++;
+                    }
+                }
+            }
+            return this;
+        }
+
+        /**
+         * Adds the other count object into this one by adding the base quality counts.
+         */
+        Counts merge(final Counts counts2) {
+            for (int i = 0; i < counts2.qCounts.length; i++) {
```

Comment:
Implicit here is that qCounts and oqCounts have the same length for both classes. I'd prefer either using a constant (like NUM_COUNT_BINS) or comping the two lengths directly.